image/unpacker.go

```diff
@@ -0,0 +1,238 @@
+// Copyright 2016 The Linux Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package image
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/opencontainers/image-spec/schema"
+	"github.com/pkg/errors"
+)
+
+// Unpacker is a structure that encapsulates the image layout unpacking
+// functionality.
+type Unpacker struct {
+	// If set, the ownership of files, directories and symblic links defined in
+	// the layers, will be preserved in the unpacked root file system.
+	PreserveOwnership bool
```

Comment:
I'd like to note that actually the implementation of a `--same-owner` flag is harder than you might think, and it's not just a matter of `chown`ing all of the files to the same owner. I don't really mind how the mapping of users is done (though I feel like the functional approach will end up being overkill when what you actually want is to have explicit `rspec.LinuxIDMapping` with an option to ignore errors [meaning to default to root] during the mapping).

If you're an unprivileged user (which is the goal of this PR) you have a bunch of restrictions that mean you have to hack around things like files which give you no read permission even though you are their owner.